keymgmt: update ECX, LMS & SLH-DSA to use generated param decoders

paulidale opened the pull request:
These three are the easy ones to convert.  The remaining ones aren't easy and will be dealt with separately.
ECX still needs some cleanup for a legacy path.

paulidale committed:
key management: rename key management files in anticipations of generated decoding

paulidale committed:
slh-dsa: convert to using a generated decoder

paulidale committed:
lms: convert to using generated parameter decoding

paulidale committed:
ecx: convert to using generated parameter decoding

slontis commented on crypto/ec/ecx_backend.c:
```diff
@@ -52,25 +52,20 @@ int ossl_ecx_public_from_private(ECX_KEY *key)
-        param_priv_key =
-            OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PRIV_KEY);
-
     if (param_pub_key == NULL && param_priv_key == NULL)
```
This logic doesnt look quite the same to me...
In the old code
````
if (param_pub_key == NULL && param_priv_key == NULL)
````
only set param_priv_key if include_private was set... So your test is different..

To be equivalent param_priv_key is passed in with include_private=1, or param_priv_key==NULL and include_private = 0, which means you either dont need include_private to be passed anymore or the above test is not right?

paulidale replied:
Is the suggestion here to drop one of the arguments since both `param_priv_key` and `include_private` aren't required?  I tried to not change how this was used, but it would be straightforward to drop the `include_private` argument.

As for a behaviour change, the additional check on line 73 below retains the existing behaviour I believe.  That's the only block that references the private key and it's now conditional on `include_private`.  The reason behind this is that we need to honour the users choice if they supplied both the public and private keys but only asked for the public part.  The generated param decoder doesn't know what the user wants and would fill point fields in if they are supplied.

paulidale replied:
Discussed this and it's okay as it is.

paulidale committed:
keymgmt: update template

The template file is exhibiting some bitrot:
1. convert to using generated param name decoders
2. add the missing `template_imexport_types` function.

paulidale committed:
changes: add a CHANGES.md entry covering the generated parameter name decoding

This isn't a user visible change but it's sufficiently noteworthy to
justify a changes entry.

t8m commented on providers/implementations/kem/ecx_kem.c.in:
```diff
@@ -143,7 +143,7 @@ static ECX_KEY *ecxkey_pubfromdata(PROV_ECX_CTX *ctx,
     if (ecx == NULL)
         return NULL;
-    if (ossl_ecx_key_fromdata(ecx, params, 0) <= 0) {
+    if (ossl_ecx_key_fromdata(ecx, params, NULL, 0) <= 0) {
```
Should this function be rewritten to make it clear the params is just a single OSSL_PARAM? Otherwise this is confusing, although it will work.

paulidale replied:
Agreed, a good suggestion.

paulidale committed:
fixup! ecx: convert to using generated parameter decoding

sshedi commented:
Hi, does this issue affect 3.5.x LTS releases? Is there any 3.5.x compatible patch for this same?

mattcaswell commented:
> Hi, does this issue affect 3.5.x LTS releases? Is there any 3.5.x compatible patch for this same?

This PR is just an internal refactor and doesn't "fix" anything per se. It was not backported to 3.5.
